### PR TITLE
cmd, dashboard, eth, p2p: separate full nodes from light ones on the dashboard

### PR DESCRIPTION
**Pull request by kurkomisi**
Supersedes https://github.com/ethereum/go-ethereum/pull/19397

This PR contains the separation of the ETH peers from the LES peers on the dashboard.

Screenshot from 2019-05-21 19-02-34

### CODE REVIEW

File: dashboard/assets/components/Dashboard.jsx [hunk 1]
```diff
@@ -83,8 +83,9 @@ const appender = <T>(limit: number, mapper = replacer) => (update: Array<T>, pre
 // the execution of unnecessary operations (e.g. copy of the log array).
 const defaultContent: () => Content = () => ({
 	general: {
-		version: null,
-		commit:  null,
+		version:  null,
+		commit:   null,
+		syncMode: '',
```
Comment (karalabe) [inline dashboard/assets/components/Dashboard.jsx]: Pls remove the sync mode, that's mostly an internal detail, is not that useful on the outside, at least not now.

File: dashboard/dashboard.go [hunk 2]
```diff
@@ -27,13 +27,17 @@ package dashboard
...
 	"net"
 	"net/http"
 	"sync"
 	"sync/atomic"
-	"time"
 
-	"io"
+	"github.com/ethereum/go-ethereum/eth"
+	"github.com/ethereum/go-ethereum/les"
+
```
Comment (karalabe) [inline dashboard/dashboard.go]: Pls remove empty lines from in between these imports. The only empty line should be between stdlib and externals.

File: dashboard/dashboard.go [hunk 3]
```diff
@@ -67,6 +71,14 @@ type Dashboard struct {
 
 	quit chan chan error // Channel used for graceful exit
 	wg   sync.WaitGroup  // Wait group used to close the data collector threads
+
+	peerCh  chan p2p.MeteredPeerEvent // Peer event channel.
+	subPeer event.Subscription        // Peer event subscription.
+
+	syncMode downloader.SyncMode
```
Comment (karalabe) [inline dashboard/dashboard.go]: Pls remove the sync mode, that's mostly an internal detail, is not that useful on the outside, at least not now.

File: dashboard/message.go [hunk 4]
```diff
@@ -37,8 +37,9 @@ type ChartEntry struct {
 }
 
 type GeneralMessage struct {
-	Version string `json:"version,omitempty"`
-	Commit  string `json:"commit,omitempty"`
+	Version  string `json:"version,omitempty"`
+	Commit   string `json:"commit,omitempty"`
+	SyncMode string `json:"syncMode,omitempty"`
```
Comment (karalabe) [inline dashboard/message.go]: Pls remove the sync mode, that's mostly an internal detail, is not that useful on the outside, at least not now.

File: eth/backend.go [hunk 5]
```diff
@@ -481,6 +481,7 @@ func (s *Ethereum) EthVersion() int                    { return int(s.protocolMa
 func (s *Ethereum) NetVersion() uint64                 { return s.networkID }
 func (s *Ethereum) Downloader() *downloader.Downloader { return s.protocolManager.downloader }
 func (s *Ethereum) Synced() bool                       { return atomic.LoadUint32(&s.protocolManager.acceptTxs) == 1 }
+func (s *Ethereum) IsLesServer() bool                  { return s.lesServer != nil }
```
Comment (karalabe) [inline eth/backend.go]: I'd rather we don't add a method here. If you use `admin.nodeInfo` (or the equivalent call against the Eth object), you should be able to derive this whether `eth`, `les` or both protocols are advertised.

File: p2p/metrics.go [hunk 6]
```diff
@@ -58,24 +55,25 @@ var (
...
 )
 
 // MeteredPeerEvent is an event emitted when peers connect or disconnect.
 type MeteredPeerEvent struct {
 	Type    MeteredPeerEventType // Type of peer event
-	IP      net.IP               // IP address of the peer
-	ID      enode.ID             // NodeID of the peer
+	Addr    string               // TCP address of the peer
 	Elapsed time.Duration        // Time elapsed between the connection and the handshake/disconnection
+	Enode   string               // Node URL
```
Comment (karalabe) [inline p2p/metrics.go]: This seems to be `Peer.Enode()`, can't we get rid of this extra redundancy?

File: p2p/server.go [hunk 7]
```diff
@@ -873,9 +876,13 @@ func (srv *Server) listenLoop() {
 			continue
 		}
 		if remoteIP != nil {
-			fd = newMeteredConn(fd, true, remoteIP)
+			var addr *net.TCPAddr
```
Comment (kurkomisi) [inline p2p/server.go]: @fjl Could you please take a look? I didn't know what is the proper solution of this merge conflict.

Comment (kurkomisi) [reply p2p/server.go]: The relevant information is that I use the `TCPAddr` instead of the `IP`.

File: p2p/metrics.go [hunk 8]
```diff
@@ -58,24 +56,24 @@ var (
...
+	// PeerDisconnected is the type of event emitted when a peer disconnects.
+	PeerDisconnected
 )
 
 // MeteredPeerEvent is an event emitted when peers connect or disconnect.
 type MeteredPeerEvent struct {
 	Type    MeteredPeerEventType // Type of peer event
-	IP      net.IP               // IP address of the peer
-	ID      enode.ID             // NodeID of the peer
+	Addr    string               // TCP address of the peer
```
Comment (karalabe) [inline p2p/metrics.go]: Isn't this also available via Peer?

**Comment by fjl** [comment]
closing because I am merging #19397 instead.

**Comment by karalabe** [comment]
Why? This is a lot more complete

**Comment by fjl** [comment]
Sorry, I meant I'm merging #19762 which includes this PR.

**Pull request closed without merging**